### PR TITLE
avoiding memory leaks on errors in weekly schedule for #1860

### LINKED ISSUE #1860: Correct WeeklySchedule to avoid memory leak and Windows incompatibility
The `WeeklySchedule` C implementation has two issues that need to be corrected:

- it won't work for files with Windows file endings (https://github.com/lbl-srg/modelica-buildings/issues/3745)
- it leaks memory (https://github.com/lbl-srg/modelica-buildings/issues/3752)

@Mathadon : What is your availability to fix these two issues? (I am working an a release for Buildings and it would be great to include this.)

### PR DESCRIPTION
**Pull request by Mathadon**
@mwetter this should fix the memory leak issue. I was able to run `IBPSA.Utilities.IO.Files.Examples.WeeklySchedule` using  this code but I did no memory checks using valgrind.

I don't have a machine to test the windows line endings. Is that something you could have a look at?

### CODE REVIEW

File: IBPSA/Utilities/IO/Files/WeeklySchedule.mo [hunk 1]
```diff
@@ -54,6 +54,10 @@ protected
         revisions="<html>
 <ul>
 <li>
+March 30 2024, by Filip Jorissen:<br/>
+Avoiding memory leaks for #1860.
```
Comment (mwetter) [inline IBPSA/Utilities/IO/Files/WeeklySchedule.mo:58]: The number should be a hyperlink to the issue.

**Comment by mwetter** [comment]
I could test the windows line endings, but can you please update the code. I am not familiar enough with the logic of your C implementation, by maybe where you test for `\n`, you could check if there is a `\r` followed by an `\n` and if so, skip the `\r`. Or alternatively convert the buffer from `\r\n` to `\n` before all the processing starts?

We should then replicate one of the validation and use a text file with Windows line endings, which we should be able to add with dos2unix.

File: IBPSA/Resources/C-Sources/WeeklySchedule.c [hunk 2]
```diff
@@ -144,105 +162,140 @@ void* weeklyScheduleInit(const int tableOnFile, const char* name, const double t
...
           ncharsRow = strcspn(source, ",");
 
           if (tokenLen == ncharsRow + 5 ) {
+            weeklyScheduleFreeInit(scheduleID);
+            weeklyScheduleFree(scheduleID);
             ModelicaFormatError("Incorrect header when reading weekly schedule '%s'. No comma was found in the header.", name);
           }
-          strncpy(buff2, source, ncharsRow);
-          buff2[ncharsRow] = '\0';
+          strncpy(scheduleID->buff2, source, ncharsRow);
```
[the remark this answers is not in the document]
Comment (Mathadon) [reply IBPSA/Resources/C-Sources/WeeklySchedule.c:179]: Good point!

[the remark this answers is not in the document]
Comment (Mathadon) [reply IBPSA/Resources/C-Sources/WeeklySchedule.c:179]: I added an explicit check although the check ` if (index >= bufLen - 2) {` might suffice.

**Review by Mathadon — COMMENTED** [review]
I have made some revisions but can't figure out from the code why it crashes in Dymola. On OM it works ok:
image


I'm not sure how to debug further since we do not have a dymola license. You also checked the simulation tab for any error messages?

**Comment by Mathadon** [comment]
@mwetter through 0ecdc2e8813cab7caf2eb330b9a3c1fa2c3bd43c I attempted to add `\r\n` support.

**Comment by mwetter** [comment]
@Mathadon : I have not checked for the root cause of the error, but also OCT refuses the translate the file. It issues:
```
free(): double free detected in tcache 2
```

**Comment by Mathadon** [comment]
@mwetter thanks for checking further! I got the same error and apparently I forgot to remove the original `fclose()`. See [ac9c6e2](https://github.com/ibpsa/modelica-ibpsa/pull/1861/commits/ac9c6e2302c831c4ebcbaf6997e2fc7424d49bc8) for a fix. Does that resolve the windows issue too?

**Comment by mwetter** [comment]
@Mathadon : Now the results are wrong:
image

**Comment by Mathadon** [comment]
@mwetter thanks for the windows example model. I found the bug and tested the new example. The results visually match the old results in your screenshot and the new example works. :) see [a911c6a](https://github.com/ibpsa/modelica-ibpsa/pull/1861/commits/a911c6ad6418dbdc80e2f0bc278439e472c66aeb)

**Comment by Mathadon** [comment]
🙌

**Pull request merged**